### PR TITLE
memberships: remove owner, member and child group

### PR DESCRIPTION
**Pull request by DTabidze**

### CODE REVIEW

File: core/auth/memberships/main.go [hunk 1]
```diff
@@ -681,6 +698,53 @@ func (s *Server) groupHandler(w http.ResponseWriter, r *http.Request) {
...
+			return exists
+		}
+		if !groupExists(parentGroup) {
+			http.Error(w, fmt.Sprintf("group with the name '%s' does not exist", parentGroup), http.StatusNotFound)
+			return
+		}
+		if !groupExists(childGroup) {
+			http.Error(w, fmt.Sprintf("group with the name '%s' does not exist", childGroup), http.StatusNotFound)
+			return
+		}
```
Comment (giolekva) [inline core/auth/memberships/main.go:734]: Is this part needed? If any of the groups does not exist, pair of (parent, child) won't exist either so store.RemoveFromGroupToGroup invocation will fail as well.

File: core/auth/memberships/group.html [hunk 2]
```diff
@@ -79,11 +80,18 @@ <h3>Child Groups</h3>
...
             <th>Description</th>
+            <th>Action</th>
         </tr>
         {{- range .ChildGroups }}
         <tr>
             <td><a href="/group/{{ .Name }}">{{ .Name }}</a></td>
             <td>{{ .Description }}</td>
+            <td>
+                <form action="/remove-child-group/{{ $parentGroupName }}/{{ .Name }}" method="post" onsubmit="return confirm('Are you sure you want to remove this group as child?')">
+                    <input type="hidden" name="child-group" value="{{ .Name }}">
```
Comment (giolekva) [inline core/auth/memberships/group.html:91]: I think this is not used

File: core/auth/memberships/group.html [hunk 3]
```diff
@@ -79,11 +80,18 @@ <h3>Child Groups</h3>
...
             <th>Group Name</th>
             <th>Description</th>
+            <th>Action</th>
         </tr>
         {{- range .ChildGroups }}
         <tr>
             <td><a href="/group/{{ .Name }}">{{ .Name }}</a></td>
             <td>{{ .Description }}</td>
+            <td>
+                <form action="/remove-child-group/{{ $parentGroupName }}/{{ .Name }}" method="post" onsubmit="return confirm('Are you sure you want to remove this group as child?')">
```
Comment (giolekva) [inline core/auth/memberships/group.html:90]: Let's do same for owners and members as well.

**Review by giolekva — CHANGES_REQUESTED** [review]
nice

File: core/auth/memberships/group.html [hunk 4]
```diff
@@ -45,7 +45,12 @@ <h4>Owners</h4>
         {{- range .Owners }}
         <tr>
             <td><a href="/user/{{ . }}">{{ . }}</a></td>
-            <td>Delete</td>
+            <td>
+                <form action="/remove-user-group/{{ . }}/{{ $parentGroupName }}" method="post" onsubmit="return confirm('Are you sure you want to remove this user from group?')">
+                    <input type="hidden" name="table-name" value="owners">
```
Comment (giolekva) [inline core/auth/memberships/group.html:50]: API should be clean, and table name is an internal detail of our current implementation which should not be expose externally.

It is cleaner and more secure to have separate endpoints `/remove-member` and `/remove-owner`

File: core/auth/memberships/group.html [hunk 5]
```diff
@@ -45,7 +45,12 @@ <h4>Owners</h4>
         {{- range .Owners }}
         <tr>
             <td><a href="/user/{{ . }}">{{ . }}</a></td>
-            <td>Delete</td>
+            <td>
+                <form action="/remove-user-group/{{ . }}/{{ $parentGroupName }}" method="post" onsubmit="return confirm('Are you sure you want to remove this user from group?')">
```
Comment (giolekva) [inline core/auth/memberships/group.html:49]: Here you are revoking ownership of the group, while error message talks about just removing user from the group.
Change message to `Are you sure you want to revoke user {{ მომხმარებელი }} ownership of the {{ ჯგუფის სახელი }} group?`

File: core/auth/memberships/group.html [hunk 6]
```diff
@@ -58,7 +63,12 @@ <h4>Members</h4>
         {{- range .Members }}
         <tr>
             <td><a href="/user/{{ . }}">{{ . }}</a></td>
-            <td>Delete</td>
+            <td>
+                <form action="/remove-user-group/{{ . }}/{{ $parentGroupName }}" method="post" onsubmit="return confirm('Are you sure you want to remove this user from group?')">
```
Comment (giolekva) [inline core/auth/memberships/group.html:67]: Include user and group names in confirmation dialog

File: core/auth/memberships/group.html [hunk 7]
```diff
@@ -58,7 +63,12 @@ <h4>Members</h4>
...
+                    <input type="hidden" name="table-name" value="user_to_group">
```
Comment (giolekva) [inline core/auth/memberships/group.html:68]: remove table-name from here as well.

**Review by giolekva — CHANGES_REQUESTED** [review]

File: core/auth/memberships/group.html (hunk 5 above)
Comment (DTabidze) [reply core/auth/memberships/group.html]: Done.

File: core/auth/memberships/group.html (hunk 4 above)
Comment (DTabidze) [reply core/auth/memberships/group.html]: Done.

File: core/auth/memberships/group.html (hunk 6 above)
Comment (DTabidze) [reply core/auth/memberships/group.html]: Done.

File: core/auth/memberships/group.html (hunk 7 above)
Comment (DTabidze) [reply core/auth/memberships/group.html]: Done.

File: core/auth/memberships/group.html [hunk 8]
```diff
@@ -44,7 +45,11 @@ <h4>Owners</h4>
         {{- range .Owners }}
         <tr>
             <td><a href="/user/{{ . }}">{{ . }}</a></td>
-            <td>Delete</td>
+            <td>
+                <form action="/remove-group-owner/{{ . }}/{{ $parentGroupName }}" method="post" onsubmit="return confirm('Are you sure you want to revoke user {{ . }} ownership of the {{ $parentGroupName }} group?')">
```
Comment (giolekva) [inline core/auth/memberships/group.html:49]: Logically group name should come first and then the owner

File: core/auth/memberships/group.html [hunk 9]
```diff
@@ -57,7 +62,11 @@ <h4>Members</h4>
         {{- range .Members }}
         <tr>
             <td><a href="/user/{{ . }}">{{ . }}</a></td>
-            <td>Delete</td>
+            <td>
+                <form action="/remove-group-member/{{ . }}/{{ $parentGroupName }}" method="post" onsubmit="return confirm('Are you sure you want to remove user {{ . }} user from {{ $parentGroupName }} group?')">
```
Comment (giolekva) [inline core/auth/memberships/group.html:66]: Same here, put group name first in the url

File: core/auth/memberships/main.go [hunk 10]
```diff
@@ -527,7 +527,7 @@ func (s *Server) Start() error {
 		r := mux.NewRouter()
 		r.PathPrefix("/static/").Handler(http.FileServer(http.FS(staticResources)))
 		r.HandleFunc("/remove-child-group/{parent-group}/{child-group}", s.removeChildGroupHandler)
-		r.HandleFunc("/remove-user-group/{username}/{groupname}", s.removeUserFromGroupHandler)
+		r.HandleFunc("/{action}/{username}/{group-name}", s.removeUserFromGroupHandler)
```
Comment (giolekva) [inline core/auth/memberships/main.go:530]: Theoretically this has a chance of matching other unintended url. If you really want to reuse the handler for these two endpoints, `/remove-{action}/{username}/{group-name}` is safer to use

**Review by giolekva — CHANGES_REQUESTED** [review]

File: core/auth/memberships/group.html (hunk 8 above)
Comment (DTabidze) [reply core/auth/memberships/group.html]: Done.

File: core/auth/memberships/group.html (hunk 9 above)
Comment (DTabidze) [reply core/auth/memberships/group.html]: Done.

File: core/auth/memberships/main.go (hunk 10 above)
Comment (DTabidze) [reply core/auth/memberships/main.go]: Done.

**Review by giolekva — APPROVED** [review]

**Pull request merged**